atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ */
+
+package com.palantir.common.compression;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.xxhash.StreamingXXHash32;
+import net.jpountz.xxhash.XXHashFactory;
+
+public class LZ4CompressingInputStream extends InputStream {
+    private final InputStream delegate;
+    private final LZ4Compressor compressor;
+    private final byte[] buffer;
+    private final byte[] compressedBuffer;
+    private final StreamingXXHash32 hasher;
+    private final LZ4FrameDescriptor frameDescriptor;
+    private int position;
+    private int maxPosition;
+    private boolean inFrameFooter;
+
+    public LZ4CompressingInputStream(InputStream delegate) throws IOException {
+        this(delegate, LZ4Streams.DEFAULT_FRAME_DESCRIPTOR);
+    }
+
+    public LZ4CompressingInputStream(InputStream delegate, LZ4FrameDescriptor frameDescriptor) throws IOException {
+        this.delegate = delegate;
+        this.compressor = LZ4Streams.compressor;
+        this.buffer = LZ4Streams.newUncompressedBuffer(frameDescriptor);
+        this.compressedBuffer = LZ4Streams.newCompressedBuffer(frameDescriptor);
+        this.hasher = XXHashFactory.fastestInstance().newStreamingHash32(0);
+        this.frameDescriptor = frameDescriptor;
+        this.position = 0;
+        this.maxPosition = LZ4Streams.FRAME_HEADER_LENGTH;
+        this.inFrameFooter = false;
+        writeFrameHeader();
+    }
+
+    @Override
+    public int read() throws IOException {
+        if ((position == maxPosition) && !refill()) {
+            return -1;
+        }
+        return compressedBuffer[position++] & 0xff;
+    }
+
+    @Override
+    public int read(byte b[], int off, int len) throws IOException {
+        if (b == null) {
+            throw new NullPointerException();
+        } else if (off < 0 || len < 0 || len > b.length - off) {
+            throw new IndexOutOfBoundsException();
+        } else if (len == 0) {
+            return 0;
+        }
+
+        int bytesRead = 0;
+        while (bytesRead < len) {
+            int remainingBuffer = maxPosition - position;
+            int bytesToRead = Math.min(len - bytesRead, remainingBuffer);
+            System.arraycopy(compressedBuffer, position, b, off + bytesRead, bytesToRead);
+            position += bytesToRead;
+            bytesRead += bytesToRead;
+            if ((position == maxPosition) && !refill()) {
+                break;
+            }
+        }
+        return bytesRead > 0 ? bytesRead : -1;
```

Comment:
General style nit: leave a blank line after a long (>1 line) if/while/for/etc block.